File: bin/cnndraw.rb
#!/usr/bin/env ruby

exit unless ARGV.length == 3

words = File.new ARGV[0]
labels = Hash.new
i = 1

words.each_line do |line|
	word = line.split ","
	labels[i.to_s] = word[0]
	i = i + 1
end

draw = ""
x = 150
y = 300
separation = 50
bar_width = 150
bar_height = 16
STDIN.each_line do |line|
	print line
	args = line.split " "
	break if args[0] == 'elapsed'
	for i in 0..(args.length / 2 - 1)
        frac_bar_width = (args[i * 2 + 1].to_f * bar_width).to_i + x
        draw += sprintf("-fill none -strokewidth 1 -stroke DodgerBlue -draw \"rectangle %d,%d,%d,%d\" ", x, y, bar_width+x, y + bar_height)
        draw += sprintf("-fill DodgerBlue -draw \"rectangle %d,%d,%d,%d\" -strokewidth 0 -stroke none ", x, y, frac_bar_width, y + bar_height)
        draw += sprintf("-fill white -pointsize 36 -annotate +%d+%d \"%s\" ", x, y , labels[args[i * 2]])
		#draw += sprintf("-fill none -strokewidth 1 -stroke DodgerBlue -draw \"rectangle 15,%d,165,%d\" -fill DodgerBlue -draw \"rectangle 15,%d,%d,%d\" -strokewidth 0 -stroke none -fill white -annotate 'text 18,%d \"%s\"' ", y, y + 16, y, (args[i * 2 + 1].to_f * 150).to_i + 15, y + 16, y + 13, labels[args[i * 2]])
		y += separation
	end
end
print draw
%x[#{sprintf("convert %s %s%s", ARGV[1], draw, ARGV[2])}]

File: bin/cnndraw.py
#!/usr/bin/python
import sys, os
sysargs = sys.argv[1:]

# -----------------------------------------
# draws the results of cnn classification
# over the image, listing the guesses with confidence values


if len(sysargs) < 3:
	quit()

words = open(sysargs[0],"r+")
labels = {}
i = 1

for line in words:
	word = line.split(',')
	labels[str(i)] = word[0]
	i = i + 1


draw = ""
y = 15
#STDIN.each_line do |line|
while True:
	line = raw_input("input: ")
	print line
	args = line.split(' ')
	if args[0] == 'elapsed':
		break
	for jj in range(len(args) / 2):
		draw += "-fill none -strokewidth 1 -stroke DodgerBlue -draw \"rectangle 15,%d,165,%d\" -fill DodgerBlue -draw \"rectangle 15,%d,%d,%d\" -strokewidth 0 -stroke none -fill red -draw 'text 18,%d \"%s\"' " % (y, y + 16, y, (int(float(args[jj*2+1])*150)) + 15, y + 16, y + 13, labels[args[jj * 2]])
		y += 31

execstr = "convert %s %s%s" % (sysargs[1], draw, sysargs[2])
print("executing \""+str(execstr)+"\"")
os.system(execstr)
